File: built-in plugins dsp/mb-creative-harmonics.cpp
/**
 * MB Harmonics Generator
 * Category : effect
 * Type     : distortion
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Generate odd and even harmonics for warmth and presence
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_CREATIVE_HARMONICS_H
#define MB_CREATIVE_HARMONICS_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbCreativeHarmonics : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-creative-harmonics";
    static constexpr const char* PLUGIN_NAME    = "MB Harmonics Generator";
    static constexpr const char* PLUGIN_TYPE    = "distortion";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float oddHarmonics = 0.3f;  // range [0, 1]
    float evenHarmonics = 0.3f;  // range [0, 1]
    float order = 5f;  // range [2, 12]
    float mix = 0.5f;  // range [0, 1]
    };

    MbCreativeHarmonics() = default;
    ~MbCreativeHarmonics() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.oddHarmonics = std::clamp(params.oddHarmonics, 0f, 1f);
        params.evenHarmonics = std::clamp(params.evenHarmonics, 0f, 1f);
        params.order = std::clamp(params.order, 2f, 12f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Harmonics Generator
        return input;
    }
};

#endif // MB_CREATIVE_HARMONICS_H

File: built-in plugins dsp/mb-creative-harmonics.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbCreativeHarmonicsPlugin: PluginDefinition = { id: 'mb-creative-harmonics', slug: 'mb-creative-harmonics', name: 'MB Harmonics Generator', category: 'effect', type: 'distortion' as any, version: '1.0.0', description: 'Generate odd and even harmonics for warmth and presence', author: 'Max Booster', parameters: [{ id: 'oddHarmonics', name: 'Odd Harmonics', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true }, { id: 'evenHarmonics', name: 'Even Harmonics', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true }, { id: 'order', name: 'Max Order', type: 'float', defaultValue: 5, minValue: 2, maxValue: 12, automatable: false }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { oddHarmonics: 0.3, evenHarmonics: 0.3, order: 5, mix: 0.5 } };

export default MbCreativeHarmonicsPlugin;
